pkg/controller/dnsendpoint/nameserver/gcp.go

```diff
@@ -194,7 +195,35 @@ func (q *gcpQuery) queryNameServer(gcpClient gcpclient.Client, managedZone strin
 
 // createNameServers creates the name servers for the specified domain in the specified managed zone.
 func (q *gcpQuery) createNameServers(gcpClient gcpclient.Client, managedZone string, domain string, values sets.String) error {
-	return gcpClient.AddResourceRecordSet(managedZone, q.resourceRecordSet(domain, values))
+	// First check whether there is an existing NS record
```

Comment:
I would like to see attempt the create first and only query for existing records if the create fails with a status that indicates that the records already exist. See Delete for a similar example.

I expect that update is a lot more rare than insert, so I want to avoid performing a query every time when it is usually unneeded. I am a bit concerned about why the DNSEndpoint controller needed to update the NS records n restart.

Reply:
the code falls through https://github.com/openshift/hive/blob/v1/pkg/controller/dnsendpoint/dnsendpoint_controller.go#L172 which leads to the Create() call

Reply:
Right, which it should not be. It should get caught by the case right above that where the current nameservers are equal to the desired nameservers. Since it is not, there is a bug. We can create a Jira card to take that bug on separate from this PR, if you think that the bug is not caused by the changes from this PR.

Reply:
I tried running this on the v1 branch, and created a GCP dnszone, and then restarted the controller and see the controller spin as it keeps trying to re-create the NS entries.

```
INFO[0005] reconciling dns endpoint                      controller=dnsendpoint dnszone=testzoneg namespace=hive
INFO[0005] creating NS record                            controller=dnsendpoint dnszone=testzoneg domain=test.jdomain.gcp.hive.openshift.com namespace=hive
ERRO[0006] error creating NS record                      controller=dnsendpoint dnszone=testzoneg domain=test.jdomain.gcp.hive.openshift.com error="error creatin
g the name server: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'test.jdomain.gcp.hive.openshift.com. (NS)' already exists, alreadyExist
s" namespace=hive
INFO[0006] reconcile complete                            controller=dnsendpoint dnszone=testzoneg domain=test.jdomain.gcp.hive.openshift.com elapsed=413.038206ms
 namespace=hive
INFO[0008] reconciling dns endpoint                      controller=dnsendpoint dnszone=testzoneg namespace=hive
INFO[0008] creating NS record                            controller=dnsendpoint dnszone=testzoneg domain=test.jdomain.gcp.hive.openshift.com namespace=hive
ERRO[0008] error creating NS record                      controller=dnsendpoint dnszone=testzoneg domain=test.jdomain.gcp.hive.openshift.com error="error creatin
g the name server: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'test.jdomain.gcp.hive.openshift.com. (NS)' already exists, alreadyExist
s" namespace=hive
INFO[0008] reconcile complete                            controller=dnsendpoint dnszone=testzoneg domain=test.jdomain.gcp.hive.openshift.com elapsed=394.647008ms
 namespace=hive
```
So this doesn't appear to be new behavior. I gave a quick glance at the code and it appears to be due to this line needing to run before a re-reconcile can skip the Create() path
https://github.com/openshift/hive/blob/v1/pkg/controller/dnsendpoint/dnsendpoint_controller.go#L178

Reply:
I created https://issues.redhat.com/browse/CO-696 to track this.

Reply:
It appears that the gpc nameserver scaper is just plain broken.